src/Mod/Draft/draftguitools/gui_edit_draft_objects.py

```diff
@@ -103,15 +104,23 @@ def update_object_from_edit_points(self, obj, node_idx, v, alt_edit_mode=0):
         pts[node_idx] = v
         obj.Points = pts
 
+    def get_edit_point_context_menu(self, edit_command, obj, node_idx):
+        return {
+            "delete point": lambda: self.handle_delete_point(edit_command, obj, node_idx),
+        }
 
-    def get_edit_point_context_menu(self, obj, node_idx):
-        actions = ["delete point"]
-        return actions
-    
-    def evaluate_context_menu_action(self, edit_command, obj, node_idx, action):
-        if action == "delete point":
-            self.delete_point(obj, node_idx)
-            edit_command.resetTrackers(obj)
+    def get_edit_obj_context_menu(self, edit_command, obj, position):
+        return {
+            "add point": lambda: self.handle_add_point(edit_command, obj, position),
+        }
+
+    def handle_delete_point(self, edit_command, obj, node_idx):
+        self.delete_point(obj, node_idx)
+        edit_command.resetTrackers(obj)
```

Comment:
It seems that *all* handle_something methods you defined call `edit_command.resetTrackers(obj)`. If this is really something that should always happen after running a command, maybe it can be called in edit_gui.py before calling the action callback (that probably means you'll have to create a wrapper around the callback, since evaluate_menu_action no longer has access to the object). Then maybe passing `edit_command` to `get_edit_point_context_menu` is also no longer necessary?

Or can you imagine a situation where calling `resetTrackers` after an action is not needed (or even problematic)?

Reply:
I was thinking about that. 
1. No I can't imagine any scenario where `resetTrackers `is not called after editing, but I didn't know even what a wrapper is :-(
2. I think passing the edit_command to `get_edit_point_context_menu `is desirable because it enables you to do that: https://github.com/yorikvanhavre/BIM_Workbench/blob/db0017f00e6e538d421551e36b06f6e9e07f4a62/archobjects/wall.py#L1138 : you can use a menu action to start an alternative editing mode. That is done also with arcs and circles.

Reply:
> but I don't know even what a wrapper is :-(

A wrapper function is a simple function that replaces another function, and calls the original function and adds something to it. That probably doesn't really help, so let me say it in code:

```python
        for (label, callback) in actions:
            action = self.tracker_menu.addAction(label)
            def wrapper(callback=callback):
                callback()
                self.resetTrackers(obj)
            action.setData(wrapper)
```

This defines a new `wrapper` function, that calls the original callback and then calls resetTrackers. Note that this creates a new function for every loop iteration, so each of these wrapper functions captures potentially different callback, self and obj values so things work as expected. Note I did something weird with the `callback` value there: Since functions like these capture a variable, not its value at the time of function definition, and loop variables like `label` and `callback` are a single variable shared between all loop iterations, capturing `callback` directly ends up with all wrappers calling the last callback (i.e. they all capture the same variable and by the time the wrappers are called, that variable will contain the last of the callbacks). This is commonly solved by using a default value in the function definition, since such a default value uses the *value* of the (in this case) `callback` variable, not capturing the variable.
Maybe I'm only making things more confusing, though...

One other thing I realized, when you posted the link to the BIM workbench, is that we can't really *change* the behavior completely as you did now, since there are third-party gui tools that would need to be updated separately. So I guess we'll need to keep compatibility with the old way of having a separate `get_edit_point_context_menu` and ` evaluate_context_menu_action`. This is not too hard, though, since you can just create callback functions that just call ` evaluate_context_menu_action` and then the rest of the code can use those. Maybe something like:

```python
          for item in actions:
              # Compatibility with the old way of specifying actoins
              if isinstance(item, str):
                  label = item
  
                  def wrapper(label=label):
                      obj_gui_tools. evaluate_context_menu_action(self, obj, ep, label)
              else:
                  (label, callback) = item
  
                  def wrapper(callback=callback):
                      callback()
                      self.resetTrackers(obj)
  
              action = self.tracker_menu.addAction(label)
              action.setData(wrapper)

```

If you add a compatibility mode, this gives even another option for more commit separation, because then you can modify display_tracker_menu first in one commit, and then modfiy all GuiTools subclasses in a next commit.

> I think passing the edit_command to get_edit_point_context_menu is desirable because it enables you to do that: https://github.com/yorikvanhavre/BIM_Workbench/blob/db0017f00e6e538d421551e36b06f6e9e07f4a62/archobjects/wall.py#L1138 : you can use a menu action to start an alternative editing mode. That is done also with arcs and circles.

Ah, indeed. So passing `edit_command` makes sense, even when it's not needed for the `resetTrackers`.

Reply:
I think i'll be reading it again tomorrow, but it is less or more clear, thanks for the deep explanation.
That code is really elegant and I'm really in favour of keeping the resetTracker call inside main Draft_Edit module.
From the commit history point of view this is for sure good, but I don't think we have to care about backward compatibility in this case: I'm 99% sure that i'm the only one that implemented the support for other objects outside Draft WB. And those BIM Tools are marked as experimental anyway.

Reply:
> but I don't think we have to care about backward compatibility in this case: 

Ok, but that does mean that the experimental tools in the BIM module are going to break when this is merged, and then break in older FC versions when BIM is updated. But that might be acceptable, especially if it's only these experimental modules, of course.

Reply:
> BIM module are going to break when this is merged
I'm going to fix it as soon as it's merged, since i can directly merge my commits in BIM wb master :)
It would be a problem for older versions, but we are talking about an experimental feature marked as unstable for production. (also if I can assure it's super usable)